src/igvfd/audit/sample.py

```diff
@@ -35,3 +35,33 @@ def audit_sample_sorted_fraction_parent_child_check(value, system):
         )
         if prop_errors != '':
             yield AuditFailure('inconsistent sorted fraction metadata', detail, level='ERROR')
+
+
+@audit_checker('Sample', frame='object')
+def audit_sample_virtual_donor_check(value, system):
+    '''Non-virtual samples should not be linked to virtual donors.'''
+    if 'virtual' and 'donors' in value:
```

Comment:
if the sample is virtual - we should skip the check - it is wasteful to traverse donors if we already know the sample is virtual.